### PR TITLE
Reducing AMQP ReactorSender memory consumption

### LINKED ISSUE #26372: Reducing AMQP ReactorSender memory consumption
The following sample code uses EH Producer API to send N events every second.

<details><summary>sample_code</summary>

```java
void recurringSend(final int eventsPerRecur, final int totalDurationInSec) {

    final EventHubProducerAsyncClient producer = new EventHubClientBuilder()
        .connectionString(System.getenv("EH_CON_STR"), System.getenv("EH_NAME"))
        .buildAsyncProducerClient();

    final ScheduledExecutorService recurringScheduler = Executors.newScheduledThreadPool(1);
    final AtomicInteger msgId = new AtomicInteger(0);

    recurringScheduler.scheduleAtFixedRate(() -> {
        List<CompletableFuture<?>> sendFutures = new ArrayList<>();

        for (int i = 0; i < eventsPerRecur; i++) {
            final byte[] data = ("msg#" + msgId.getAndIncrement()).getBytes(StandardCharsets.UTF_8);

            final Mono<Void> sendMono = producer.createBatch().flatMap(batch -> {
                batch.tryAdd(new EventData(data));
                return producer.send(batch);
            });

            final CompletableFuture<Void> sendFuture = sendMono.toFuture();
            sendFutures.add(sendFuture);
        }

        try {
            final CompletableFuture<Void> mergedFuture
                = CompletableFuture.allOf(sendFutures.toArray(CompletableFuture[]::new));
            mergedFuture.get();
        } catch (InterruptedException | ExecutionException e) {
            System.err.println("Error occurred while waiting for the result. " + e);
        }
    }, 1, 1, TimeUnit.SECONDS);

    try {
        recurringScheduler.awaitTermination(totalDurationInSec, TimeUnit.SECONDS);
        recurringScheduler.shutdown();
    } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
    } finally {
        producer.close();
    }
}
```

</details>

Below is the memory snapshot for this code execution with eventsPerRecur:1000 and totalDurationInSec:960. The memory peeked to ~2 GB and then consistently stayed around ~1.5GB. 

The setup - EventHubs and VM is running on the same region, using the latest (at the time of testing) azure-core-amqp version: 2.3.5. 

Sender_1_5GB

After digging into the profiler data and trying out different approaches, it turns out that using `Mono.defer()` instead of `Mono.just()` in two places in "Mono<Integer>"ReactorSender::getLinkSize()" method brought down the memory usage, with peek now capped at 700 MB.

i.e., peek memory usage reduced by a factor of 2.5 (2Gb -> 700Mb).

Sender_700MB

_Note that the purpose of this doc is not to promise that 700Mb (or any number) is the expected memory limit but only to describe the changes that reduced the overall memory consumption_.

There might be other potential areas where we can save memory usage, but need to discover those with proper profiling to avoid code change that causes regression.

### PR DESCRIPTION
**Pull request by anuchandy**
Refer to the git ticket https://github.com/Azure/azure-sdk-for-java/issues/26372 for details.

Fixes #26372

### CODE REVIEW

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java [hunk 1]
```diff
@@ -317,12 +317,13 @@ public String getHostname() {
     @Override
     public Mono<Integer> getLinkSize() {
         if (linkSize > 0) {
-            return Mono.just(this.linkSize);
+            // Using Mono.defer when returning cached linkSize, refer https://github.com/Azure/azure-sdk-for-java/issues/26372.
+            return Mono.defer(() -> Mono.just(this.linkSize));
```
Comment (alzimmermsft) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java:321]: When the `Mono.defer` is subscribed to does it use the `linkSize` captured during the construction of the lambda or the value at the time of subscription?

Comment (anuchandy) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java:321]: The lambda object captures `this` instance not the properties of `this`. The properties are read when lambda executes.

For example -

```java
private volatile int i = 0; 
public void foo() {     
    Mono<Integer> m = Mono.defer(() -> Mono.just(this.i));     
    this.i = 10;    
    System.out.println(m.block());  // outputs 10    
    this.i = 20;     
    System.out.println(m.block());  // outputs 20
 }
```

Side note: Also, in the ReactorSender, once linkSize is read from the link only once, it won't change once set.

**Review by conniey — APPROVED** [review]

**Review by alzimmermsft — APPROVED** [review]

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java (hunk 1 above)
Comment (anuchandy) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java:334]: Capturing the offline discussion for future reference - for example, when we evaluate the potential places to replace _just_ with _defer_.

Why the code change is safe in this specific case -

```java
if (this.linkSize > 0) {   // Imperative check for cache value
    return Mono.defer(() -> Mono.just(this.linkSize));  // Reactive access to cached value
}
```

1. The `linkSize` can be set only once, so its value will not change between the Imperative check and Reactive (delayed) access.
2. The lambda in `Mono.defer` will not capture the `linkSize` property; it gets eagerly read (as a result of subscription).
3. The `linkSize` is volatile, hence accessing the cached value from a different thread (e.g., Project Reactor thread) will see the latest value (and won't get partial value).

**Pull request merged**